### PR TITLE
Hadoop-BAM and htsjdk upgrades, Spark tool changes.

### PR DESCRIPTION
**Pull request by cmnbroad**
Requires upgrades to Hadoop-BAM (not yet released) and htsjdk.

Fixes https://github.com/broadinstitute/gatk/issues/1346, https://github.com/broadinstitute/gatk/issues/1261, https://github.com/broadinstitute/gatk/issues/1175, https://github.com/broadinstitute/gatk/issues/1326, https://github.com/broadinstitute/gatk/issues/1259, and much of the underlying code for https://github.com/broadinstitute/gatk/issues/1270, which will be enabled in a separate PR.

### CODE REVIEW

**Comment by droazen** [comment]
Ping me once this is ready

**Comment by cmnbroad** [comment]
@droazen Hadoop-BAM has been released now so this is ready for review now whenever you can get to it. There are two other changes included that are unrelated to the tickets listed above:
* ValidateSAMFile has a change because the setValidateIndex API has been deprecated in htsjdk
* Includes a centralized DEFAULT_READ_VALIDATION_STRINGENCY value per our discussion.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java [hunk 1]
```diff
@@ -313,9 +313,9 @@ protected void runPipeline( JavaSparkContext sparkContext ) {
     /**
      * Initialize standard tool inputs.
      */
-    private void initializeToolInputs(JavaSparkContext sparkContext) {
-        initializeReads(sparkContext);
+    private void initializeToolInputs(final JavaSparkContext sparkContext) {
         initializeReference();
+        initializeReads(sparkContext); // reference must be intialized before reads
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java]: `intialized` -> `initialized`

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 2]
```diff
@@ -42,8 +46,16 @@
     private static final String HADOOP_PART_PREFIX = "part-";
 
     private transient final JavaSparkContext ctx;
-    public ReadsSparkSource(JavaSparkContext ctx) {
+    private ValidationStringency validationStringency = ReadConstants.DEFAULT_READ_VALIDATION_STRINGENCY;
+
+    protected final Logger logger = LogManager.getLogger(ReadsSparkSource.class);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: We generally make our loggers `static`

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 3]
```diff
@@ -69,40 +81,24 @@ public ReadsSparkSource(JavaSparkContext ctx) {
      *                  use the default split size (determined by the Hadoop input format, typically the size of one HDFS block).
      * @return RDD of (SAMRecord-backed) GATKReads from the file.
      */
-    public JavaRDD<GATKRead> getParallelReads(final String readFileName, final String referencePath, final List<SimpleInterval> intervals, long splitSize) {
-        final Configuration conf = new Configuration();
+    public JavaRDD<GATKRead> getParallelReads(final String readFileName, final String referencePath, final List<SimpleInterval> intervals, final long splitSize) {
+        final Configuration conf = ctx.hadoopConfiguration();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Add a comment explaining why we need to get the `Configuration` from `ctx` rather than default-constructing a new one.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 4]
```diff
@@ -201,6 +199,41 @@ public boolean accept(Path path) {
     }
 
     /**
+     * Propagate any values that need to be passed to Hadoop-BAM through configuration properties:
+     *
+     *   - the validation stringency property is always set
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Mention that it's set to the current value in the `validationStringency` field

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 5]
```diff
@@ -201,6 +199,41 @@ public boolean accept(Path path) {
...
+     *   - if the input file is a CRAM file, the reference value, which must be a URI which includes
+     *     a scheme, will also be set
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Mention that if it has no scheme, it will default to `file://`

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 6]
```diff
@@ -201,6 +199,41 @@ public boolean accept(Path path) {
...
+     *   - if the input file is not CRAM, the reference property is *unset* to prevent Hadoop-BAM
+     *     from passing a stale value through to htsjdk when multiple read calls are made serially
+     *     with different inputs but the same Spark context
+     */
+    private void setHadoopBAMConfigurationProperties(final String inputName, final String referenceName) {
+        final Configuration conf = ctx.hadoopConfiguration();
+        conf.set(SAMHeaderReader.VALIDATION_STRINGENCY_PROPERTY, validationStringency.name());
+
+        if (!IOUtils.isCramFileName(inputName)) { // only set the reference for CRAM input
+            conf.unset(CRAMInputFormat.REFERENCE_SOURCE_PATH_PROPERTY);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: This call will silently succeed if the property being unset is already unset, correct?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 5 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Also mention that the method will throw if the reference is null and the input is a CRAM

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 7]
```diff
@@ -201,6 +199,41 @@ public boolean accept(Path path) {
...
+        }
+        else {
+            if (null == referenceName) {
+                throw new UserException.MissingReference("A reference is required for CRAM input");
+            }
+            else {
+                if (ReferenceTwoBitSource.isTwoBit(referenceName)) { // htsjdk can't handle 2bit reference files
+                    throw new UserException("A 2bit file cannot be used as a CRAM file reference");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Can you create a ticket to address this limitation?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 8]
```diff
@@ -69,40 +81,24 @@ public ReadsSparkSource(JavaSparkContext ctx) {
...
-            conf.set(CRAMInputFormat.REFERENCE_SOURCE_PATH_PROPERTY, referencePath);
+        setHadoopBAMConfigurationProperties(readFileName, referencePath);
 
-            rdd2 = ctx.newAPIHadoopFile(
-                    readFileName, CRAMInputFormat.class, LongWritable.class, SAMRecordWritable.class,
-                    conf);
-        } else {
-            rdd2 = ctx.newAPIHadoopFile(
+        rdd2 = ctx.newAPIHadoopFile(
                     readFileName, AnySAMInputFormat.class, LongWritable.class, SAMRecordWritable.class,
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Are we no longer using `CRAMInputFormat` for crams? Can you explain why this is no longer necessary?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 9]
```diff
@@ -161,15 +157,16 @@ public ReadsSparkSource(JavaSparkContext ctx) {
     /**
      * Loads the header using Hadoop-BAM.
      * @param filePath path to the bam.
+     * @param referencePath to the reference
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Document whether this can be null and under what circumstances

File: src/main/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFile.java [hunk 10]
```diff
@@ -114,7 +114,7 @@ protected Object doWork() {
             validator.setBisulfiteSequenced(IS_BISULFITE_SEQUENCED);
         }
         if (VALIDATE_INDEX) {
-            validator.setValidateIndex(VALIDATE_INDEX);
+            validator.setIndexValidationStringency(BamIndexValidator.IndexValidationStringency.EXHAUSTIVE);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFile.java]: Is this a change from the previous index validation stringency level?

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 11]
```diff
@@ -33,14 +36,13 @@
     @DataProvider(name = "loadReads")
     public Object[][] loadReads() {
         return new Object[][]{
-                {NA12878_chr17_1k_BAM, null},
+                {NA12878_chr17_1k_BAM, null, ValidationStringency.SILENT},
 
-                {dirBQSR + "HiSeq.1mb.1RG.2k_lines.alternate.bam",  null},
-                {dirBQSR + "expected.HiSeq.1mb.1RG.2k_lines.alternate.recalibrated.DIQ.bam", null},
+                {dirBQSR + "HiSeq.1mb.1RG.2k_lines.alternate.bam", null, ValidationStringency.STRICT},
+                {dirBQSR + "expected.HiSeq.1mb.1RG.2k_lines.alternate.recalibrated.DIQ.bam", null, ValidationStringency.STRICT},
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: Is there a reason why some test cases are using `SILENT` and others are using `STRICT`?

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 12]
```diff
@@ -89,13 +118,23 @@ public void testHeadersAreStripped() {
...
+    }
+
     @Test
     public void testPartitionSizing(){
 
         String bam = dirBQSR + "HiSeq.1mb.1RG.2k_lines.alternate.bam"; //file is ~220 kB
         JavaSparkContext ctx = SparkContextFactory.getTestSparkContext();
 
-        ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        ReadsSparkSource readSource = new ReadsSparkSource(ctx, ValidationStringency.STRICT);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: I think that all of these tests (except the ones testing the effect of validation stringency itself) should use the one-arg `ReadsSparkSource` constructor (to use the default validation stringency) instead of passing in `STRICT` or `SILENT` explicitly. It's better to test that we can pass these tests with the default stringency (whatever it may be) rather than burdening test authors with having to tailor the stringency to each input.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 13]
```diff
@@ -125,7 +164,39 @@ public void testReadFromFileAndHDFS() throws IOException {
...
+        final File reference = new File(v37_chr17_1Mb_Reference);
+        final File referenceIndex = new File(v37_chr17_1Mb_Reference + ".fai");
+
+        MiniDFSCluster cluster = null;
+        try {
+            cluster = new MiniDFSCluster.Builder(new Configuration()).build();
+            final Path workingDirectory = cluster.getFileSystem().getWorkingDirectory();
+            final Path cramPath = new Path(workingDirectory,"hdfs.cram");
+            final Path refPath = new Path(workingDirectory, "hdfs.fasta");
+            final Path refIndexPath = new Path(workingDirectory, "hdfs.fasta.fai");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: Name these `cramHDFSPath`, `refHDFSPath`, and `refIndexHDFSPath` for clarity

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 14]
```diff
@@ -65,12 +74,32 @@ public void readsSparkSourceTest(String bam, String referencePath) {
...
+    }
+
     @Test(dataProvider = "loadShardedReads", groups = "spark")
     public void shardedReadsSparkSourceTest(String expectedBam, String shardedBam, String referencePath) {
         JavaSparkContext ctx = SparkContextFactory.getTestSparkContext();
 
-        ReadsSparkSource readSource = new ReadsSparkSource(ctx);
-        JavaRDD<GATKRead> rddSerialReads = getSerialReads(ctx, expectedBam, referencePath);
+        ReadsSparkSource readSource = new ReadsSparkSource(ctx, ValidationStringency.STRICT);
+        JavaRDD<GATKRead> rddSerialReads = getSerialReads(ctx, expectedBam, referencePath, ValidationStringency.DEFAULT_STRINGENCY);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: Should this be the default set in `ReadConstants`?

**Comment by droazen** [comment]
Review complete -- back to @cmnbroad

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 3 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: I'm going to revert this change as its actually no longer necessary for my changes. Originally, I was setting the reference and stringency properties here, and Hadoop-BAM didn't see them if I wrote them to the newly constructed conf (they have to be set on the context's conf). However, I'm now setting those properties in setHadoopBAMConfigurationProperties, which uses the context conf.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 8 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: AnySAMInputFormat delegates to the appropriate input format; originally it didn't know about CRAMInputFormat but now it does so we no longer need to branch on that  here.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 3 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: There's no need to revert if you think using the `Configuration` from the `ctx

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 6 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Yes, though I'm saying that based on empirical evidence. The doc doesn't specify the behavior.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 7 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: It already exists https://github.com/broadinstitute/gatk/issues/1443.

File: src/main/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFile.java (hunk 10 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFile.java]: AFAIK, index validation was binary before, and the concept of stringency level didn't exist. It was introduced by https://github.com/samtools/htsjdk/pull/397. I think the idea was to introduce a less exhaustive and faster validation level. So I think exhaustive is the equivalent of what we used to do. Whether we want to relax that is another question. See https://github.com/broadinstitute/picard/issues/346.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 3 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Right, but I think we (at least *I*) don't. I'm reluctant to change it if its not necessary - I'm unclear on how it affects the splitsize code below. Also see [this exchange](https://broad.hipchat.com/history/room/1020897/2016/01/20?p=1&q=configuration&t=rid-1020897#8d8de87d-4c43-4e80-b588-8467017e1935) from a few weeks back. I don't understand how the newHadoopAPI method uses the two configs it has access to (one attached to the context  object and one passed explicitly as an arg).

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: I've also been continually confused by the context/confs and what is necessary from them.  If it's necessary that we always use the same one, we should be much more careful about threading it through everywhere that it is needed.  If it's not, can we just always default construct them?  

I suspect that what is important is that we are consistently using the same settings everywhere we need to, and we should commit to either ignoring the local environment by always constructing a new one, or always thread the same one through everywhere.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: This is something we should really figure out -- perhaps @tomwhite would know?

Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Hadoop configuration can be tricky... Constructing a new configuration is generally not a good idea as you get a default one, without any of the settings you (or other parts of the system) have applied. So we should try to use the existing configuration (from the context).

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java (hunk 14 above)
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: Yes. Done.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 3 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Since I see that we have a test for the partition size code, which gives me more confidence in this change, I'm going to leave it in, with the result that we use the context configuration everywhere (in this file anyway) and never create a new one.

**Comment by cmnbroad** [comment]
I think I addressed everything. Back to @droazen.

**Comment by droazen** [comment]
:+1: merge when ready
